pkg/syncer/util.go

```diff
@@ -623,3 +625,89 @@ func getPVsInBoundAvailableOrReleasedForVc(ctx context.Context, metadataSyncer *
 
 	return k8svolumes, nil
 }
+
+// createVolumeOnMultiVc attemts to create a static volume on each VC until it gets SUCCESS.
+// If while creating the volume, CNS returns CnsAlreadyRegisteredFault, it means that the volume does not need to be re-created.
+func createVolumeOnMultiVc(ctx context.Context, pv *v1.PersistentVolume,
+	metadataSyncer *metadataSyncInformer, volumeType string, metadataList []cnstypes.BaseCnsEntityMetadata, volumeHandle string) (string, volumes.Manager, error) {
+	log := logger.GetLogger(ctx)
+
+	vcconfigs, err := cnsvsphere.GetVirtualCenterConfigs(ctx, metadataSyncer.configInfo.Cfg)
+	if err != nil {
+		return "", nil, logger.LogNewErrorf(log, "failed to get VirtualCenterConfigs. err: %v", err)
+	}
+
+	for _, vcconfig := range vcconfigs {
+		log.Debugf("Attempting to create volume on VC %s", vcconfig.Host)
+		volManager, err := getVolManagerForVcHost(ctx, vcconfig.Host, metadataSyncer)
+		if err != nil {
+			continue
+		}
+		err = createCnsVolume(ctx, pv, metadataSyncer, volManager, volumeType, vcconfig.Host, metadataList, volumeHandle)
+		if err == nil {
+			return vcconfig.Host, volManager, nil
+		}
+		log.Debugf("Failed to create volume %s on VC %s", volumeHandle, vcconfig.Host)
+	}
+	return "", nil, logger.LogNewErrorf(log,
+		"Failed to create volume %s on any of the VCs", volumeHandle)
+}
+
+func createCnsVolume(ctx context.Context, pv *v1.PersistentVolume,
+	metadataSyncer *metadataSyncInformer, cnsVolumeMgr volumes.Manager, volumeType string, vcHost string, metadataList []cnstypes.BaseCnsEntityMetadata, volumeHandle string) error {
+	log := logger.GetLogger(ctx)
+
+	vcHostObj, vcHostObjFound := metadataSyncer.configInfo.Cfg.VirtualCenter[vcHost]
+	if !vcHostObjFound {
+		return logger.LogNewErrorf(log,
+			"Failed to find VC host for given volume: %q.", volumeHandle)
+	}
+
+	containerCluster := cnsvsphere.GetContainerCluster(clusterIDforVolumeMetadata,
+		vcHostObj.User, metadataSyncer.clusterFlavor,
+		metadataSyncer.configInfo.Cfg.Global.ClusterDistribution)
+
+	createSpec := &cnstypes.CnsVolumeCreateSpec{
+		Name:       pv.Name,
+		VolumeType: volumeType,
+		Metadata: cnstypes.CnsVolumeMetadata{
+			ContainerCluster:      containerCluster,
+			ContainerClusterArray: []cnstypes.CnsContainerCluster{containerCluster},
+			EntityMetadata:        metadataList,
+		},
+	}
+
+	if volumeType == common.BlockVolumeType {
+		createSpec.BackingObjectDetails = &cnstypes.CnsBlockBackingDetails{
```

Comment:
if this is for registering file volume as CNS volume, then why are we tacking care of setting block volume type?

Reply:
This method is not just for block volume. It applies to file volume also. Since both of file and block had a lot of common code, I combined it into this method. They're being called with appropriate arguments wherever necessary.